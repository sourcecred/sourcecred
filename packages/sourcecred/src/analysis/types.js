// @flow

import {
  type NodeAddressT,
  type EdgeAddressT,
  NodeAddress,
  EdgeAddress,
} from "../core/graph";
import {type EdgeWeight, edgeWeightParser} from "../core/weights";
import * as G from "../util/combo";

/**
 * This module defines `NodeType`s and `EdgeType`s, both of which are
 * data structures containing shared metadata that describes many nodes or edges
 * in the graph. Nodes can be "members" of zero or more `NodeType`s, and edges can
 * be "members" of zero or more `EdgeType`s. Membership is determined based on the
 * type's `prefix`, which is an address. A node or edge is considered a member of
 * a type if that type's prefix is a prefix of the node's address.
 *
 * To make this more concrete, let's consider a specific example. Suppose we define
 * the following nodes:
 *
 * const pullNode = NodeAddress.fromParts(["github", "pull", "repo", "34"]);
 * const commitNode = NodeAddress.fromParts(["git", "commit", "e1337"]);
 * const pullType: NodeType = {
 *  name: "Pull request",
 *  prefix: NodeAddress.fromParts(["github", "pull"]),
 *  // ... more properties as required
 * };
 * const githubType: NodeType = {
 *   name: "GitHub node",
 *   prefix: NodeAddress.fromParts(["github"])
 * };
 *
 * Then the `pullNode` is considered a member of the `pullType` and `githubType`,
 * while the `commitNode` is not a member of either type.
 *
 * The concept of a "type" is useful to SourceCred because it let's us express
 * that groups of nodes are conceptually related, and that we should treat them similarly.
 * Most saliently, we use types to assign default weights to groups of nodes, and to
 * aggregate them for better UI organization. The fact that the SourceCred UI can group
 * all pull requests together, and assign a default weight to all of them, is possible
 * because the GitHub plugin defines a pull request node type.
 *
 * While a node or edge can theoretically be a member of multiple types, in practice we
 * generally treat the node or edge as though it is only a member of its most specific type.
 * In the example above, we would treat any individual pull request as though it is only
 * a member of the pull request node type. That may change in the future. In general,
 * the type system is not wholly finalized; when it does become finalized, we will
 * likely move it into src/core. See [#710] for context.
 *
 * [#710]: https://github.com/sourcecred/sourcecred/issues/710
 */

/**
 * Represents a "Type" of node in the graph. See the module docstring
 * for context.
 */
export type NodeType = {|
  // The name for an individual node of this type.
  // For example, for the GitHub PULL_REQUEST node type, the name is "Pull
  // request". The first letter of the name should be capitalized.
  +name: string,
  // The name for multiple nodes of this type.
  // For example, for the GitHub PULL_REQUEST node type, the pluralName is
  // "Pull requests". The first letter of this name should be capitalized.
  +pluralName: string,
  // The address that will be used to test whether a node is a member
  // of this NodeType. A given node `n` is a member of the type `t` if
  // `NodeAddress.hasPrefix(n, t.prefix) == true`
  +prefix: NodeAddressT,
  // The default weight to assign to nodes of this type. We use `1` to mean "of
  // normal importance", and the weights scale linearly from there (i.e. 2
  // means twice as important).
  +defaultWeight: number,
  // The `description` property should be a human-readable string that makes
  // it clear to a user what each NodeType does
  +description: string,
|};

export const nodeTypeParser: G.Parser<NodeType> = G.object({
  name: G.string,
  pluralName: G.string,
  prefix: NodeAddress.parser,
  defaultWeight: G.number,
  description: G.string,
});

/**
 * Represents a "Type" of edge in the graph. See the module docstring
 * for context.
 */
export type EdgeType = {|
  // A brief descriptive name of what the "forward" direction of the edge
  // means. For example, for the GitHub REFERENCES edge type, the forwardName
  // is "references"
  +forwardName: string,

  // A brief descriptive name of what the "backward" direction of the edge
  // means. For example, for the GitHub REFERENCES edge type, the backwardName
  // is "referenced by"
  +backwardName: string,

  // The default forwards and backwards weights for this edge.
  // We use `1` as a default value ("of normal importance").
  // The weights have linear importance, i.e. 2 is twice as important as 1.
  +defaultWeight: EdgeWeight,

  // The address that will be used to test whether an edge is a member
  // of this EdgeType. A given edge `e` is a member of the type `t` if
  // `EdgeAddress.hasPrefix(e.address, t.prefix) == true`
  +prefix: EdgeAddressT,

  // The `description` property should be a human-readable string that makes
  // it clear to a user what each EdgeType does.
  //
  // By convention, the first line of the description should be a sentence
  // beginning with the word "Connects", which describes what kinds of nodes
  // the edge connects and why. Optionally, you may provide examples and additional
  // context after one blank link. Here is an example (for the "Merged As" edge).
  //
  // ```js
  // const mergedAsDescription = dedent`\
  //  Connects a GitHub pull request to the Git commit it merged.
  //
  //  A pull request can have either one or zero Merged As edges, depending on
  //  whether or not it was ever merged.
  //  `
  // ```
  //
  // (Note the use of the `util/dedent.js` makes it easier to write multi-lined
  // strings with clean formatting.)
  +description: string,
|};

export const edgeTypeParser: G.Parser<EdgeType> = G.object({
  forwardName: G.string,
  backwardName: G.string,
  defaultWeight: edgeWeightParser,
  prefix: EdgeAddress.parser,
  description: G.string,
});

export type NodeAndEdgeTypes = {|
  +nodeTypes: NodeType[],
  +edgeTypes: EdgeType[],
|};
