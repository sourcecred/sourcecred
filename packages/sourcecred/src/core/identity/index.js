// @flow

export type {IdentityId} from "./id";
export {identityIdParser, delimitedIdentityIdParser} from "./id";
export type {Identity} from "./identity";
export {
  newIdentity,
  contractions,
  graphNode,
  parser as identityParser,
  IDENTITY_PREFIX,
} from "./identity";

export type {IdentityType} from "./identityType";
export {parser as identityTypeParser} from "./identityType";

export type {Name} from "./name";
export {nameFromString, parser as nameParser} from "./name";

export type {Alias} from "./alias";
export {parser as aliasParser} from "./alias";
