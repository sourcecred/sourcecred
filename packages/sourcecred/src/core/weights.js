// @flow

import deepEqual from "lodash.isequal";
import * as MapUtil from "../util/map";
import * as C from "../util/combo";
import {
  type NodeAddressT,
  type EdgeAddressT,
  NodeAddress,
  EdgeAddress,
} from "../core/graph";
import {toCompat, type Compatible, compatibleParser} from "../util/compat";

/**
 * Represents the weight for a particular Node (or node address prefix).
 * Weight 1 is the default value and signifies normal importance.
 * Weights are linear, so 2 is twice as important as 1.
 */
export type NodeWeight = number;

export type NodeOperator = (NodeWeight, NodeWeight) => NodeWeight;

/**
 * Represents the forwards and backwards weights for a particular Edge (or
 * edge address prefix).
 * Weight 1 is the default value and signifies normal importance.
 * Weights are linear, so 2 is twice as important as 1.
 */
export type EdgeWeight = {|+forwards: number, +backwards: number|};
export const edgeWeightParser: C.Parser<EdgeWeight> = C.object({
  forwards: C.number,
  backwards: C.number,
});

export type EdgeOperator = (EdgeWeight, EdgeWeight) => EdgeWeight;

/**
 * Represents the weights for nodes and edges.
 *
 * The weights are stored by address prefix, i.e. multiple weights may apply
 * to a given node or edge.
 */
export type WeightsT = {|
  nodeWeights: Map<NodeAddressT, NodeWeight>,
  // Map from an edge prefix or address to a weight
  edgeWeights: Map<EdgeAddressT, EdgeWeight>,
|};

/**
 * Creates new, empty weights.
 */
export function empty(): WeightsT {
  return {
    nodeWeights: new Map(),
    edgeWeights: new Map(),
  };
}

export function copy(w: WeightsT): WeightsT {
  return {
    nodeWeights: new Map(w.nodeWeights),
    edgeWeights: new Map(w.edgeWeights),
  };
}

/** Merge multiple Weights together.
 *
 * The resultant Weights will have every weight specified by each of the input
 * weights.
 *
 * When there are overlaps (i.e. the same address is present in two or more of
 * the Weights), then the appropriate resolver will be invoked to resolve the
 * conflict. The resolver takes two weights and combines them to return a new
 * weight.
 *
 * When no resolvers are explicitly provided, merge defaults to
 * conservative "error on conflict" resolvers.
 */
export function merge(
  ws: $ReadOnlyArray<WeightsT>,
  resolvers: ?{|+nodeResolver: NodeOperator, +edgeResolver: EdgeOperator|}
): WeightsT {
  if (resolvers == null) {
    const nodeResolver = (_unused_a, _unused_b) => {
      throw new Error(
        "node weight conflict detected, but no resolver specified"
      );
    };
    const edgeResolver = (_unused_a, _unused_b) => {
      throw new Error(
        "edge weight conflict detected, but no resolver specified"
      );
    };
    resolvers = {nodeResolver, edgeResolver};
  }
  const weights: WeightsT = empty();
  const {nodeWeights, edgeWeights} = weights;
  const {nodeResolver, edgeResolver} = resolvers;
  for (const w of ws) {
    for (const [addr, val] of w.nodeWeights.entries()) {
      const existing = nodeWeights.get(addr);
      if (existing == null) {
        nodeWeights.set(addr, val);
      } else {
        try {
          nodeWeights.set(addr, nodeResolver(existing, val));
        } catch (e) {
          throw new Error(`${e} when resolving ${NodeAddress.toString(addr)}`);
        }
      }
    }
    for (const [addr, val] of w.edgeWeights.entries()) {
      const existing = edgeWeights.get(addr);
      if (existing == null) {
        edgeWeights.set(addr, val);
      } else {
        try {
          edgeWeights.set(addr, edgeResolver(existing, val));
        } catch (e) {
          throw new Error(
            `Error ${e} when resolving ${EdgeAddress.toString(addr)}`
          );
        }
      }
    }
  }
  return weights;
}

export type SerializedWeights_0_2_0 = {|
  +nodeWeights: {[NodeAddressT]: NodeWeight},
  +edgeWeights: {[EdgeAddressT]: EdgeWeight},
|};

function serialize_0_2_0(weights: WeightsT): SerializedWeights_0_2_0 {
  return {
    nodeWeights: MapUtil.toObject(weights.nodeWeights),
    edgeWeights: MapUtil.toObject(weights.edgeWeights),
  };
}

function deserialize_0_2_0(weights: SerializedWeights_0_2_0): WeightsT {
  return {
    nodeWeights: MapUtil.fromObject(weights.nodeWeights),
    edgeWeights: MapUtil.fromObject(weights.edgeWeights),
  };
}

const Parse_0_2_0: C.Parser<SerializedWeights_0_2_0> = (() => {
  return C.object({
    nodeWeights: C.dict(C.number, NodeAddress.parser),
    edgeWeights: C.dict(edgeWeightParser, EdgeAddress.parser),
  });
})();

const COMPAT_INFO = {type: "sourcecred/weights", version: "0.2.0"};

export const parser: C.Parser<WeightsT> = compatibleParser(COMPAT_INFO.type, {
  "0.2.0": C.fmap(Parse_0_2_0, deserialize_0_2_0),
});

export type WeightsJSON_0_2_0 = Compatible<SerializedWeights_0_2_0>;
export type WeightsJSON = WeightsJSON_0_2_0;

export function toJSON(weights: WeightsT): WeightsJSON {
  return toCompat(COMPAT_INFO, serialize_0_2_0(weights));
}

export function fromJSON(json: WeightsJSON): WeightsT {
  return parser.parseOrThrow(json);
}

export type NodeWeightDiff = {|
  +address: NodeAddressT,
  +first: ?NodeWeight,
  +second: ?NodeWeight,
|};

export type EdgeWeightDiff = {|
  +address: EdgeAddressT,
  +first: ?EdgeWeight,
  +second: ?EdgeWeight,
|};

export type WeightsComparison = {|
  +weightsAreEqual: boolean,
  +nodeWeightDiffs: $ReadOnlyArray<NodeWeightDiff>,
  +edgeWeightDiffs: $ReadOnlyArray<EdgeWeightDiff>,
|};

export function compareWeights(
  firstWeights: WeightsT,
  secondWeights: WeightsT
): WeightsComparison {
  const nodeWeightDiffs = [];
  const edgeWeightDiffs = [];

  const nodeAddresses = new Set([
    ...firstWeights.nodeWeights.keys(),
    ...secondWeights.nodeWeights.keys(),
  ]);
  for (const address of nodeAddresses) {
    const first = firstWeights.nodeWeights.get(address);
    const second = secondWeights.nodeWeights.get(address);
    if (!deepEqual(first, second)) {
      nodeWeightDiffs.push({
        address,
        first,
        second,
      });
    }
  }
  const edgeAddresses = new Set([
    ...firstWeights.edgeWeights.keys(),
    ...secondWeights.edgeWeights.keys(),
  ]);
  for (const address of edgeAddresses) {
    const first = firstWeights.edgeWeights.get(address);
    const second = secondWeights.edgeWeights.get(address);
    if (!deepEqual(first, second)) {
      edgeWeightDiffs.push({
        address,
        first,
        second,
      });
    }
  }
  const weightsAreEqual =
    nodeWeightDiffs.length === 0 && edgeWeightDiffs.length === 0;

  return {
    weightsAreEqual,
    nodeWeightDiffs,
    edgeWeightDiffs,
  };
}
