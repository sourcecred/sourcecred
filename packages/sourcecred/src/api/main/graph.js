// @flow

import {type WeightedGraph} from "../../core/weightedGraph";
import {Ledger} from "../../core/ledger/ledger";
import type {PluginDirectoryContext} from "../plugin";
import {type PluginId} from "../pluginId";
import {Plugin} from "../plugin";
import {type TaskReporter, SilentTaskReporter} from "../../util/taskReporter";
import {CascadingReferenceDetector} from "../../core/references/cascadingReferenceDetector";
import {type ReferenceDetector} from "../../core/references/referenceDetector";
import {ensureIdentityExists} from "../../core/ledger/identityProposal";

export type GraphInput = {|
  +plugins: Array<{|
    +plugin: Plugin,
    +directoryContext: PluginDirectoryContext,
    +pluginId: PluginId,
  |}>,
  +ledger: Ledger,
|};

export type GraphOutput = {|
  +pluginGraphs: $ReadOnlyArray<{|
    +pluginId: PluginId,
    +weightedGraph: WeightedGraph,
  |}>,
  +ledger: Ledger,
|};

/**
 Iterates through the provided plugins, runs their `graph` and `identities`
 processes, and updates the ledger with any new IdentityProposals.
 
 Might mutate the ledger that is passed in.
 */
export async function graph(
  input: GraphInput,
  // Specifies which of the plugins in the GraphInput should be run.
  // If omitted, all plugins in the GraphInput are run.
  scope?: $ReadOnlyArray<PluginId>,
  taskReporter: TaskReporter = new SilentTaskReporter()
): Promise<GraphOutput> {
  // Build Reference Detector
  const rds = [];
  for (const {pluginId, plugin, directoryContext} of input.plugins) {
    const task = `reference detector for ${pluginId}`;
    taskReporter.start(task);
    const rd = await plugin.referenceDetector(directoryContext, taskReporter);
    rds.push(rd);
    taskReporter.finish(task);
  }
  rds.push(_hackyIdentityNameReferenceDetector(input.ledger));
  const rd = new CascadingReferenceDetector(rds);
  // Build graphs
  const pluginGraphs = [];
  for (const {pluginId, plugin, directoryContext} of input.plugins) {
    if (!scope || scope.includes(pluginId)) {
      const task = `generating graph for ${pluginId}`;
      taskReporter.start(task);
      pluginGraphs.push({
        pluginId: pluginId,
        weightedGraph: await plugin.graph(directoryContext, rd, taskReporter),
      });
      const identities = await plugin.identities(
        directoryContext,
        taskReporter
      );
      for (const identityProposal of identities) {
        ensureIdentityExists(input.ledger, identityProposal);
      }
      taskReporter.finish(task);
    }
  }
  return {
    pluginGraphs,
    ledger: input.ledger,
  };
}

// Hack to support old-school (deprecated) "initiatives files":
// We need to be able to parse references to usernames, e.g. "@yalor", so
// we need a reference detector that will match these to identities in the
// Ledger. This isn't a robust addressing scheme, since identities are re-nameable;
// in v2 the initiatives plugin will be re-written to use identity node addresses instead.
// This hack can be safely deleted once we no longer support initiatives files that refer
// to identities by their names instead of their IDs.
export function _hackyIdentityNameReferenceDetector(
  ledger: Ledger
): ReferenceDetector {
  const usernameToAddress = new Map(
    ledger
      .accounts()
      .map((a) => [a.identity.name.toLowerCase(), a.identity.address])
  );
  function addressFromUrl(potentialUsername: string) {
    const prepped = potentialUsername.replace("@", "").toLowerCase();
    return usernameToAddress.get(prepped);
  }
  return {addressFromUrl};
}
