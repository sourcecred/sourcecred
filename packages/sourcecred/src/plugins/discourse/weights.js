// @flow

import {type NodeWeight} from "../../core/weights";
import * as C from "../../util/combo";
import * as MapUtil from "../../util/map";
import {orElse as either} from "../../util/null";
import {DEFAULT_TRUST_LEVEL_TO_WEIGHT} from "./createGraph";
import {type User} from "./fetch";
import {type CategoryId as MirrorCategoryId, type Tag} from "./fetch";

// Expected to be an integer string, like "1" or "123"
export opaque type CategoryId: string = string;
// No restrictions (can be more specific later if we investigate Discourse)
export opaque type TagId: string = string;

export function parseCategoryId(id: string): CategoryId {
  const result = parseInt(id, 10);
  if (Number.isNaN(result) || result.toString() !== id) {
    throw new Error(`CategoryId should be a string integer; got ${id}`);
  }
  return id;
}

// parse out the opaque type
export function parseTagId(id: string): TagId {
  return id;
}

export type SerializedWeightsConfig = {|
  /**
   * Tags can be configured to confer specific like-weight multipliers when
   * added to a Topic.
   * If a tag does not have a configured weight, the defaultWeight is applied.
   * An example configuration might look like:
   * ```
   * "weights": {
   *   "defaultTagWeight": 1,
   *   "tagWeights": {
   *     "foo": 0,
   *     "bar": 1.25,
   *     "baz": 2
   *   }
   *   // categoryWeight configs...
   * }
   * ```
   * where foo and bar are the names of tags used in discourse.
   *
   * When multiple tags are assigned to a topic, their weights are multiplied
   * together to yield a total tag Weight multiplier. In our example configuration,
   * if both foo and bar are added to a topic, likes on posts in the topic will
   * have a weight of 0, (0 * 1.25 = 0), which means that no cred will be minted
   * by those likes.
   *
   * If "bar" and "baz" are both added to another topic, the likes on all posts
   * in that topic will carry a weight of 2.5 (1.25 * 2 = 2.5), which means that
   * 2.5x as much cred will be minted by those likes.
   */
  +defaultTagWeight?: NodeWeight,
  +tagWeights?: {|[TagId]: NodeWeight|},
  /**
   * Categories can be configured to confer a specific like-weight multiplier
   * when added to a Topic.
   * If a category does not have a configured weight, the defaultWeight is applied.
   * An example configuration might look like:
   * ```
   * weights: {
   *  "defaultCategoryWeight": 1,
   *  "categoryWeights": {
   *    "5": 0,
   *    "36": 1.25
   *  }
   *  // tagWeight configs...
   * }
   * ```
   * where "5" and "36" are the categoryIds in discourse.
   *
   * An easy way to find the categoryId for a given category is to browse to the
   * categories section in discourse
   * (e.g. https://discourse.sourcecred.io/categories).
   * Then mousing over or clicking on a category will bring you to a url that
   * has the shape https://exampleUrl.com/c/<category name>/<categoryId>
   * Clicking on the community category in sourcecred navigates to
   * https://discourse.sourcecred.io/c/community/26 for example, where the
   * categoryId is 26
   */
  +defaultCategoryWeight?: NodeWeight,
  +categoryWeights?: {|[CategoryId]: NodeWeight|},
  /*
   * Tags and Topic Category can be used in combination to create like-weight
   * multipliers on Topics. This means that if any assigned tag or topic category
   * is set to 0, all likes on that topic will mint zero cred.
   */
|};

export type WeightsConfig = {|
  +defaultTagWeight: number,
  +tagWeights: Map<TagId, number>,
  +defaultCategoryWeight: number,
  +categoryWeights: Map<CategoryId, number>,
|};

function upgrade(s: SerializedWeightsConfig): WeightsConfig {
  return {
    defaultTagWeight: either(s.defaultTagWeight, 1),
    defaultCategoryWeight: either(s.defaultCategoryWeight, 1),
    tagWeights: MapUtil.fromObject(s.tagWeights || {}),
    categoryWeights: MapUtil.fromObject(s.categoryWeights || {}),
  };
}

export const serializedWeightsConfigParser: C.Parser<SerializedWeightsConfig> = C.object(
  {},
  {
    defaultCategoryWeight: C.number,
    defaultTagWeight: C.number,
    categoryWeights: C.dict(
      C.number,
      C.fmap(C.delimited("//"), parseCategoryId)
    ),
    tagWeights: C.dict(C.number, C.fmap(C.string, parseTagId)),
  }
);

export const weightsConfigParser: C.Parser<WeightsConfig> = C.fmap(
  serializedWeightsConfigParser,
  upgrade
);

export function likeWeight(
  weights: WeightsConfig,
  user: ?User,
  category: ?MirrorCategoryId,
  tags: ?$ReadOnlyArray<Tag>
): NodeWeight {
  const trustLevel = user == null ? 0 : _trustLevelWeight(user.trustLevel);
  const categoryWeight =
    category == null ? 1 : _categoryWeight(category, weights);
  const tagWeight = tags == null ? 1 : _weightFromTags(tags, weights);
  return trustLevel * categoryWeight * tagWeight;
}

export function _trustLevelWeight(trustLevel: number | null): NodeWeight {
  if (trustLevel == null) {
    // The null trust level shouldn't happen in practice, right now users who
    // only like but never post will have null trust level (will be fixed by #2045).
    // This means they could have trust level 1. But to be conservative, we treat anyone
    // with a null trust level as if they have trust level 0.
    // Possibly this could come up with deleted users too.
    return DEFAULT_TRUST_LEVEL_TO_WEIGHT["0"];
  }

  const key = String(trustLevel);
  const weight = DEFAULT_TRUST_LEVEL_TO_WEIGHT[key];
  if (weight == null) {
    throw new Error(`invalid trust level: ${String(key)}`);
  }
  return weight;
}

export function _categoryWeight(
  category: MirrorCategoryId,
  {categoryWeights, defaultCategoryWeight}: WeightsConfig
): NodeWeight {
  const weight = categoryWeights.get(category);
  return weight == null ? defaultCategoryWeight : weight;
}

export function _weightFromTags(
  tags: $ReadOnlyArray<Tag>,
  {tagWeights, defaultTagWeight}: WeightsConfig
): NodeWeight {
  const weightForTag = (tagId: TagId) =>
    either(tagWeights.get(tagId), defaultTagWeight);
  return tags.reduce((acc, tag) => acc * weightForTag(tag), 1);
}
