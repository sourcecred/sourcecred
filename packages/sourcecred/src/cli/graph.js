// @flow

import stringify from "json-stable-stringify";
import {LoggingTaskReporter} from "../util/taskReporter";
import type {Command} from "./command";
import dedent from "../util/dedent";
import {
  compareGraphs,
  nodeToString,
  edgeToString,
  NodeAddress,
  EdgeAddress,
} from "../core/graph";
import {compareWeights} from "../core/weights";
import {type WeightedGraph} from "../core/weightedGraph";
import * as pluginId from "../api/pluginId";
import {Instance} from "../api/instance/instance";
import {LocalInstance} from "../api/instance/localInstance";
import {graph} from "../api/main/graph";

function die(std, message) {
  std.err("fatal: " + message);
  return 1;
}

const graphCommand: Command = async (args, std) => {
  let shouldIncludeDiff = false;
  let isSimulation = false;
  const processedArgs = args.filter((arg) => {
    switch (arg) {
      case "-d":
        shouldIncludeDiff = true;
        return false;
      case "--simulation":
      case "-s":
        isSimulation = true;
        return false;
      default:
        return true;
    }
  });

  const taskReporter = new LoggingTaskReporter();
  taskReporter.start("graph");
  const baseDir = process.cwd();
  const instance: Instance = await new LocalInstance(baseDir);
  const graphInput = await instance.readGraphInput();

  const availablePlugins = graphInput.plugins.map(({pluginId}) => pluginId);
  let pluginsToLoad = [];
  if (processedArgs.length === 0) {
    pluginsToLoad = availablePlugins;
  } else {
    for (const arg of processedArgs) {
      const id = pluginId.fromString(arg);
      if (availablePlugins.includes(id)) {
        pluginsToLoad.push(id);
      } else {
        return die(
          std,
          `can't find plugin ${id}; remember to use fully scoped name, as in sourcecred/github`
        );
      }
    }
  }

  const graphOutput = await graph(graphInput, pluginsToLoad, taskReporter);

  if (shouldIncludeDiff) {
    for (const {pluginId, weightedGraph} of graphOutput.pluginGraphs) {
      const diffTask = `${pluginId}: diffing with existing graph`;
      taskReporter.start(diffTask);
      try {
        const oldWeightedGraph = await instance.readWeightedGraphForPlugin(
          pluginId
        );
        computeAndLogDiff(oldWeightedGraph, weightedGraph, pluginId, std);
      } catch (error) {
        std.out(
          `Could not find or compare existing graph.json for ${pluginId}. ${error}`
        );
      }
      taskReporter.finish(diffTask);
    }
  }

  if (!isSimulation) {
    taskReporter.start("writing files");
    await instance.writeGraphOutput(graphOutput);
    taskReporter.finish("writing files");
  }
  taskReporter.finish("graph");
  return 0;
};

export const graphHelp: Command = async (args, std) => {
  std.out(
    dedent`\
      usage: sourcecred graph [options]

      options:
      -d  Outputs the diff of changes compared to the last saved graph.
      -s, --simulation  Skips writing changes to the graph and ledger jsons.

      Generate a graph from cached plugin data

      Either 'sourcecred load' must immediately precede this command or
      a cache directory must exist for all plugins specified in sourcecred.json
      `.trimRight()
  );
  return 0;
};

function computeAndLogDiff(
  oldWeightedGraph: WeightedGraph,
  newWeightedGraph: WeightedGraph,
  pluginName: string,
  std
) {
  const graphDiff = compareGraphs(
    oldWeightedGraph.graph,
    newWeightedGraph.graph
  );
  const weightDiff = compareWeights(
    oldWeightedGraph.weights,
    newWeightedGraph.weights
  );
  const horizontalRule =
    "=============================================================";
  if (graphDiff.nodeDiffs.length > 0) {
    std.out(
      `${horizontalRule}\n  ${pluginName} - Node Diffs\n${horizontalRule}`
    );
    for (const nodeDiff of graphDiff.nodeDiffs) {
      std.out(
        `Old:\t${
          nodeDiff.first ? nodeToString(nodeDiff.first) : "No matching address"
        }`
      );
      std.out(
        `New:\t${
          nodeDiff.second
            ? nodeToString(nodeDiff.second)
            : "No matching address"
        }\n`
      );
    }
  }
  if (weightDiff.nodeWeightDiffs.length > 0) {
    std.out(
      `${horizontalRule}\n  ${pluginName} - Node Weight Diffs\n${horizontalRule}`
    );
    for (const nodeWeightDiff of weightDiff.nodeWeightDiffs) {
      std.out(`${NodeAddress.toString(nodeWeightDiff.address)}`);
      std.out(
        `Old:\t${
          nodeWeightDiff.first != null
            ? nodeWeightDiff.first.toString()
            : "No matching address"
        }`
      );
      std.out(
        `New:\t${
          nodeWeightDiff.second != null
            ? nodeWeightDiff.second.toString()
            : "No matching address"
        }\n`
      );
    }
  }
  if (graphDiff.edgeDiffs.length > 0) {
    std.out(
      `${horizontalRule}\n  ${pluginName} - Edge Diffs\n${horizontalRule}`
    );
    for (const edgeDiff of graphDiff.edgeDiffs) {
      std.out(
        `Old:\t${
          edgeDiff.first ? edgeToString(edgeDiff.first) : "No matching address"
        }`
      );
      std.out(
        `New:\t${
          edgeDiff.second
            ? edgeToString(edgeDiff.second)
            : "No matching address"
        }\n`
      );
    }
  }
  if (weightDiff.edgeWeightDiffs.length > 0) {
    std.out(
      `${horizontalRule}\n  ${pluginName} - Edge Weight Diffs\n${horizontalRule}`
    );
    for (const edgeWeightDiff of weightDiff.edgeWeightDiffs) {
      std.out(`${EdgeAddress.toString(edgeWeightDiff.address)}`);
      std.out(
        `Old:\t${
          edgeWeightDiff.first
            ? stringify(edgeWeightDiff.first)
            : "No matching address"
        }`
      );
      std.out(
        `New:\t${
          edgeWeightDiff.second
            ? stringify(edgeWeightDiff.second)
            : "No matching address"
        }\n`
      );
    }
  }
  if (graphDiff.graphsAreEqual && weightDiff.weightsAreEqual)
    std.out(
      `${horizontalRule}\n  ${pluginName} - Unchanged\n${horizontalRule}`
    );
  else
    std.out(`${horizontalRule}\n  ${pluginName} - Summary of Changes\n${horizontalRule}
Node Diffs: ${graphDiff.nodeDiffs.length}
Node Weight Diffs: ${weightDiff.nodeWeightDiffs.length}
Edge Diffs: ${graphDiff.edgeDiffs.length}
Edge Weight Diffs: ${weightDiff.edgeWeightDiffs.length}`);
}

export default graphCommand;
