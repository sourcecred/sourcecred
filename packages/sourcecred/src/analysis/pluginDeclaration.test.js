// @flow

import stringify from "json-stable-stringify";
import deepFreeze from "deep-freeze";
import {NodeAddress, EdgeAddress} from "../core/graph";
import {type NodeType, type EdgeType} from "./types";
import {
  weightsForDeclaration,
  type PluginDeclaration,
  toJSON,
  fromJSON,
} from "./pluginDeclaration";
import * as Weights from "../core/weights";

describe("analysis/pluginDeclaration", () => {
  const nodeType: NodeType = deepFreeze({
    name: "node",
    pluralName: "nodes",
    prefix: NodeAddress.fromParts(["node"]),
    defaultWeight: 2,
    description: "a type",
  });
  const edgeType: EdgeType = deepFreeze({
    forwardName: "points",
    backwardName: "is pointed to",
    prefix: EdgeAddress.fromParts(["edge"]),
    defaultWeight: {forwards: 2, backwards: 3},
    description: "a type",
  });
  const emptyDeclaration: PluginDeclaration = deepFreeze({
    name: "empty",
    nodePrefix: NodeAddress.empty,
    edgePrefix: EdgeAddress.empty,
    nodeTypes: [],
    edgeTypes: [],
    userTypes: [],
  });
  const nonEmptyDeclaration: PluginDeclaration = deepFreeze({
    name: "non-empty",
    nodePrefix: NodeAddress.empty,
    edgePrefix: EdgeAddress.empty,
    nodeTypes: [nodeType],
    edgeTypes: [edgeType],
    userTypes: [],
  });
  describe("weightsForDeclaration", () => {
    it("works for an empty declaration", () => {
      expect(weightsForDeclaration(emptyDeclaration)).toEqual(Weights.empty());
    });
    it("works for a non-empty declaration", () => {
      const expected = Weights.empty();
      expected.nodeWeights.set(nodeType.prefix, nodeType.defaultWeight);
      expected.edgeWeights.set(edgeType.prefix, edgeType.defaultWeight);
      const actual = weightsForDeclaration(nonEmptyDeclaration);
      expect(expected).toEqual(actual);
    });
  });

  describe("to/fromJSON", () => {
    it("works round-trip on an empty declaration", () => {
      const json = toJSON(emptyDeclaration);
      const result = fromJSON(json);
      expect(result).toEqual(emptyDeclaration);
    });
    it("snapshots on an empty declaration", () => {
      // stringify to avoid having literal NUL bytes in our source.
      expect(stringify([toJSON(emptyDeclaration)])).toMatchSnapshot();
    });
    it("works round-trip on an non-empty declaration", () => {
      const json = toJSON(nonEmptyDeclaration);
      const result = fromJSON(json);
      expect(result).toEqual(nonEmptyDeclaration);
    });
    it("snapshots on an non-empty declaration", () => {
      // stringify to avoid having literal NUL bytes in our source.
      expect(stringify([toJSON(nonEmptyDeclaration)])).toMatchSnapshot();
    });
  });
});
