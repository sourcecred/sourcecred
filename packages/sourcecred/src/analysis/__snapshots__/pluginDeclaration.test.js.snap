// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analysis/pluginDeclaration to/fromJSON snapshots on an empty declaration 1`] = `"[[{\\"type\\":\\"sourcecred/pluginDeclarations\\",\\"version\\":\\"0.1.0\\"},{\\"edgePrefix\\":\\"E\\\\u0000\\",\\"edgeTypes\\":[],\\"name\\":\\"empty\\",\\"nodePrefix\\":\\"N\\\\u0000\\",\\"nodeTypes\\":[],\\"userTypes\\":[]}]]"`;

exports[`analysis/pluginDeclaration to/fromJSON snapshots on an non-empty declaration 1`] = `"[[{\\"type\\":\\"sourcecred/pluginDeclarations\\",\\"version\\":\\"0.1.0\\"},{\\"edgePrefix\\":\\"E\\\\u0000\\",\\"edgeTypes\\":[{\\"backwardName\\":\\"is pointed to\\",\\"defaultWeight\\":{\\"backwards\\":3,\\"forwards\\":2},\\"description\\":\\"a type\\",\\"forwardName\\":\\"points\\",\\"prefix\\":\\"E\\\\u0000edge\\\\u0000\\"}],\\"name\\":\\"non-empty\\",\\"nodePrefix\\":\\"N\\\\u0000\\",\\"nodeTypes\\":[{\\"defaultWeight\\":2,\\"description\\":\\"a type\\",\\"name\\":\\"node\\",\\"pluralName\\":\\"nodes\\",\\"prefix\\":\\"N\\\\u0000node\\\\u0000\\"}],\\"userTypes\\":[]}]]"`;
